// @zcloak/crypto/cjs/multibase/bs64.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isBase64 = exports.base64Validate = exports.base64Encode = exports.base64Decode = void 0;
var _base = require("@scure/base");
var _helpers = require("./helpers");
// Copyright 2021-2023 zcloak authors & contributors
// SPDX-License-Identifier: Apache-2.0

const config = {
  chars: 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/',
  coder: _base.base64,
  prefix: 'm',
  type: 'base64'
};

/**
 * @description
 * Validates that the supplied value is valid base64
 */
const base64Validate = (0, _helpers.createValidate)(config);

/**
 * @description Checks if the input is in base64, returning true/false
 */
exports.base64Validate = base64Validate;
const isBase64 = (0, _helpers.createIs)(base64Validate);

/**
 * @description
 * From the provided input, decode the base64 and return the result as an `Uint8Array`.
 */
exports.isBase64 = isBase64;
const base64Decode = (0, _helpers.createDecode)(config, base64Validate);

/**
 * @description
 * From the provided input, create the base64 and return the result as a string.
 */
exports.base64Decode = base64Decode;
const base64Encode = (0, _helpers.createEncode)(config);
exports.base64Encode = base64Encode;
