// @zcloak/crypto/cjs/ed25519/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "convertPublicKeyToCurve25519", {
  enumerable: true,
  get: function () {
    return _convertKey.convertPublicKeyToCurve25519;
  }
});
Object.defineProperty(exports, "convertSecretKeyToCurve25519", {
  enumerable: true,
  get: function () {
    return _convertKey.convertSecretKeyToCurve25519;
  }
});
Object.defineProperty(exports, "ed25519DeriveHard", {
  enumerable: true,
  get: function () {
    return _deriveHard.ed25519DeriveHard;
  }
});
Object.defineProperty(exports, "ed25519PairFromSeed", {
  enumerable: true,
  get: function () {
    return _fromSeed.ed25519PairFromSeed;
  }
});
Object.defineProperty(exports, "ed25519Sign", {
  enumerable: true,
  get: function () {
    return _sign.ed25519Sign;
  }
});
Object.defineProperty(exports, "ed25519Verify", {
  enumerable: true,
  get: function () {
    return _verify.ed25519Verify;
  }
});
var _convertKey = require("./convertKey");
var _deriveHard = require("./deriveHard");
var _fromSeed = require("./fromSeed");
var _sign = require("./sign");
var _verify = require("./verify");
